### PR TITLE
Fuzzer for Pallet Bags List

### PR DESCRIPTION
**Pull request by emostov**
This PR adds a fuzzer for the `SortedListProvider` API exposed by pallet-bags-list.

### CODE REVIEW

File: frame/bags-list/fuzzer/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+fn main() {
+	ExtBuilder::default().build_and_execute(|| loop {
+		fuzz!(|data: (AccountId, VoteWeight)| {
+			let (account_id_seed, vote_weight) = data;
+
+			let id = account_id_seed % ID_RANGE;
+
+			match account_id_seed % OPTIONS {
+				0 => {
```
Comment (kianenigma) [inline frame/bags-list/fuzzer/src/main.rs:45]: nit: you can use an enum here to specify which operation is which number. transmute will work, more options here: https://stackoverflow.com/questions/28028854/how-do-i-match-enum-values-with-an-integer/29530566

File: frame/bags-list/src/mock.rs [hunk 2]
```diff
@@ -101,12 +101,13 @@ pub(crate) const GENESIS_IDS: [(AccountId, VoteWeight); 4] =
...
 
 #[derive(Default)]
-pub(crate) struct ExtBuilder {
+pub struct ExtBuilder {
 	ids: Vec<(AccountId, VoteWeight)>,
 }
 
 impl ExtBuilder {
 	/// Add some AccountIds to insert into `List`.
+	#[cfg(not(feature = "fuzz"))]
```
Comment (kianenigma) [inline frame/bags-list/src/mock.rs:110]: hmm, why?

File: frame/bags-list/fuzzer/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+					if BagsList::on_insert(id.clone(), vote_weight).is_err() {
+						// this was a duplicate id, which is ok. We can just update it.
+						BagsList::on_update(&id, vote_weight);
+					}
+					assert!(BagsList::contains(&id));
+				},
+				1 => {
+					if !BagsList::contains(&id) {
```
Comment (kianenigma) [inline frame/bags-list/fuzzer/src/main.rs:53]: arguably we could also let the raw apis be accessed without us helping it such as "insert before update if it is not there". I think `on_update` handles non-existing ids, right?

File: frame/bags-list/fuzzer/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+					}
+					BagsList::on_update(&id, vote_weight);
+					assert!(BagsList::contains(&id));
+				},
+				_ => {
+					// only `2` should ever get us here, but we must use a wildcard to make the
+					// compiler happy.
+					BagsList::on_remove(&id);
+					assert!(!BagsList::contains(&id));
+				},
```
Comment (kianenigma) [inline frame/bags-list/fuzzer/src/main.rs:65]:
```suggestion
				2 => {
					BagsList::on_remove(&id);
					assert!(!BagsList::contains(&id));
				},
				_ => unreachable!(),
```

more rust-y

File: frame/bags-list/fuzzer/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+			}
+
+			assert!(BagsList::sanity_check().is_ok());
```
Comment (kianenigma) [inline frame/bags-list/fuzzer/src/main.rs:68]: might be a good time for you to go and re-read `sanity_check` and make sure it is checking everything there is to check.

**Review by kianenigma — APPROVED** [review]
Looks reasonable. I think our CI team already runs this stuff for a few hours per day or sth. Let's also make sure we give it a solid days/week.

**Comment by emostov** [comment]
Not sure if there is a better way to share this data, but here is stats from a 22 hour run on commit 7d43123e716a112d8ef9f0b421a2ae5b947623d3

Screen Shot 2021-09-24 at 4 08 07 PM

File: frame/bags-list/src/mock.rs (hunk 2 above)
Comment (emostov) [reply frame/bags-list/src/mock.rs:110]: It was an attempt to get the CI not break when it sees unused code .. but seems to be causing other issues in the CI now

File: frame/bags-list/fuzzer/src/main.rs (hunk 1 above)
Comment (emostov) [reply frame/bags-list/fuzzer/src/main.rs]: Nice, lmk what you think about the style of my approach - a bit more verbose then using transmute, but I rather avoid unsafe code if its not neccesary

File: frame/bags-list/fuzzer/src/main.rs (hunk 3 above)
Comment (emostov) [reply frame/bags-list/fuzzer/src/main.rs]: It handles non-existing ids by just ignoring them. I changed it to make sure we exercise that codepath. My thinking though was the faster we hit the max number of ids the better.

7111f38ca5548815f9933850bd734ec1d03b8c58

File: frame/bags-list/fuzzer/src/main.rs (hunk 5 above)
Comment (emostov) [reply frame/bags-list/fuzzer/src/main.rs:85]: I found one small check to add 7cca9059d2bfa29422ed438acb1ee01c49a5326b

**Comment by emostov** [comment]
bot merge

**Comment by thiolliere** [comment]
bot merge

**Comment by shawntabrizi** [comment]
@emostov can you post the updated status from the running process?

**Comment by emostov** [comment]
Status as of October 31st:

Screen Shot 2021-10-31 at 10 33 54 PM

**Pull request merged**